File: 5 最长回文子串.py
#给定一个字符串 s，找到 s 中最长的回文子串。你可以假设 s 的最大长度为 1000。
class Solution:
    def longestPalindrome(self, s: str) -> str:
        i=0
        ans=[]
        st=list(s)
        while(i<len(s)):
            j=0
            temp=0
            while(i-j>=0 and i+j<len(s)):
                if(st[i-j]!=st[i+j]):
                    break
                temp=j*2+1
                j+=1
            if(temp>len(ans)):
                j-=1#回退
                ans=st[i-j:i+j+1]
            j=0
            while(i-j>=0 and i+j+1<len(s)):
                if(st[i-j]!=st[i+j+1]):
                    break
                temp=j*2+2
                j+=1
            if(temp>len(ans)):
                j-=1#回退
                ans=st[i-j:i+j+2]
            j=0
            i+=1
        return "".join(ans)
            

File: 5 最长回文子串.cpp
//给定一个字符串 s，找到 s 中最长的回文子串。你可以假设 s 的最大长度为 1000。
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        vector<vector<int>> dp(n, vector<int>(n));
        string ans;
        for(int k=0;k<n;k++){//k为右移位数
            for(int i=0;i+k<n;i++){//i为子串起始位置
                int j=i+k;
                if(j>=n){
                    break;
                }
                if(k==0){
                    dp[i][j]=1;
                }
                else if(k==1){
                    dp[i][j]=(s[i] == s[j]);
                }
                else{
                    dp[i][j]=(s[i] == s[j] && dp[i + 1][j - 1]);
                }
                if(dp[i][j] &&k+1>ans.size()){
                    ans=s.substr(i, k + 1);
                }
            }
        }
        return ans;
    }
};
